### ISSUE TITLE
[Bug]: Tiling is not working correctly for some models

### ISSUE THREAD

**Issue by blaz-r** [openvinotoolkit/anomalib#1318]
### Describe the bug

Tiling is currently not working correctly for STFPM and Reverse Distillation. Both implementations assume that tiler is initialized ([STFPM](https://github.com/blaz-r/anomalib/blob/ensemble_eval/src/anomalib/models/stfpm/torch_model.py#L44-L49), [reverse distillation](https://github.com/blaz-r/anomalib/blob/ensemble_eval/src/anomalib/models/reverse_distillation/torch_model.py#L48-L53)) when model class constructor is called, but this is not the case as tiler gets initialized in callback.
This is not an issue in padim and patchcore, as those two models only tile embeddings.

This would require refactor of either how tiler is initialized or change the way constructors setup dimensions.

### Dataset

MVTec

### Model

STFPM

### Steps to reproduce the behavior

1. Install from github
2. Add following code inside [get_callbacks](https://github.com/blaz-r/anomalib/blob/ensemble_eval/src/anomalib/utils/callbacks/__init__.py#L47) to insert tiling:
```python
    tiling = TilerConfigurationCallback(
        enable=config.dataset.tiling.apply,
        tile_size=config.dataset.tiling.tile_size,
        stride=config.dataset.tiling.stride,
    )
    callbacks.append(tiling)
```
4. Train STFPM with tiler
5. Everything is successful but results are not scaled correctly (see screenshot bellow)

### OS information

OS information:
- OS: Windows 10
- Python version: 3.10
- Anomalib version: 0.7.0
- PyTorch version: 2.0.1
- CUDA/cuDNN version: 11.7
- GPU models and configuration: 1x gtx 1060
- Any other relevant information: /

### Expected behavior

Training is done and visual results are correct.

### Screenshots

As you can see, due to wrong dimensions, the mask is incorrectly up-scaled and untiled.
![image](https://github.com/openvinotoolkit/anomalib/assets/61357777/7013ea36-7871-4b9e-9433-8c64ce815f3e)
This should be:
![image](https://github.com/openvinotoolkit/anomalib/assets/61357777/395517b5-584b-419d-afde-b630930fcb11)

### Pip/GitHub

GitHub

### What version/branch did you use?

latest

### Configuration YAML

```yaml
dataset:
  name: mvtec
  format: mvtec
  path: ./datasets/MVTec
  category: bottle
  task: segmentation
  train_batch_size: 32
  eval_batch_size: 32
  inference_batch_size: 32
  num_workers: 0
  image_size: 256 # dimensions to which images are resized (mandatory)
  center_crop: null # dimensions to which images are center-cropped after resizing (optional)
  normalization: imagenet # data distribution to which the images will be normalized: [none, imagenet]
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)
  tiling:
    apply: true
    tile_size: 128
    stride: null
    remove_border_count: 0
    use_random_tiling: False
    random_tile_count: 16

model:
  name: stfpm
  backbone: resnet18
  layers:
    - layer1
    - layer2
    - layer3
  lr: 0.4
  momentum: 0.9
  weight_decay: 0.0001
  early_stopping:
    patience: 5
    metric: pixel_AUROC
    mode: max
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 0
  path: ./results

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: torch, onnx, openvino
# PL Trainer Args. Don't add extra parameter here.
trainer:
  enable_checkpointing: true
  default_root_dir: null
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1 # Don't validate before extracting features.
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 100
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0 # Don't validate before extracting features.
  log_every_n_steps: 50
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle
```

### Logs

```shell
...
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
